src/Compilers/CSharp/Test/Emit/CodeGen/IndexAndRangeTests.cs

```diff
@@ -4236,5 +4236,968 @@ static void M(Span<int> s)
+    }
+}
+""";
+            var comp = CreateCompilationWithIndex(source);
+            comp.VerifyDiagnostics(
+                // (1,31): error CS1918: Members of property 'Buffer10Container.this[int]' of type 'Buffer10' cannot be assigned with an object initializer because it is of a value type
+                // _ = new Buffer10Container() { [1] = { [2] = 42 } }; // 1
+                Diagnostic(ErrorCode.ERR_ValueTypePropertyInObjectInitializer, "[1]").WithArguments("Buffer10Container.this[int]", "Buffer10").WithLocation(1, 31),
+                // (2,31): error CS1918: Members of property 'Buffer10Container.this[int]' of type 'Buffer10' cannot be assigned with an object initializer because it is of a value type
+                // _ = new Buffer10Container() { [^1] = { [^2] = 42 } }; // 2
+                Diagnostic(ErrorCode.ERR_ValueTypePropertyInObjectInitializer, "[^1]").WithArguments("Buffer10Container.this[int]", "Buffer10").WithLocation(2, 31)
+                );
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/67533")]
+        public void InObjectInitializer_Nested_Struct_WriteOnly()
+        {
+            string source = """
+_ = new Buffer10Container() { [1] = { [2] = 42 } }; // 1
+_ = new Buffer10Container() { [^1] = { [^2] = 42 } }; // 2
+
+struct Buffer10Container
+{
+    public Buffer10Container() { }
+
+    public int Length { get { System.Console.Write("ContainerLength "); return 10; } }
+    public Buffer10 this[int x]
+    {
+        set => throw null;
+    }
+}
+
+struct Buffer10
+{
+    public Buffer10() { }
+
+    public int Length { get { System.Console.Write("Length "); return 10; } }
+    public int this[int x]
+    {
+        set => throw null;
+    }
+}
+""";
+            var comp = CreateCompilationWithIndex(source);
+            comp.VerifyDiagnostics(
+                // (1,31): error CS1918: Members of property 'Buffer10Container.this[int]' of type 'Buffer10' cannot be assigned with an object initializer because it is of a value type
+                // _ = new Buffer10Container() { [1] = { [2] = 42 } }; // 1
+                Diagnostic(ErrorCode.ERR_ValueTypePropertyInObjectInitializer, "[1]").WithArguments("Buffer10Container.this[int]", "Buffer10").WithLocation(1, 31),
+                // (2,31): error CS1918: Members of property 'Buffer10Container.this[int]' of type 'Buffer10' cannot be assigned with an object initializer because it is of a value type
+                // _ = new Buffer10Container() { [^1] = { [^2] = 42 } }; // 2
+                Diagnostic(ErrorCode.ERR_ValueTypePropertyInObjectInitializer, "[^1]").WithArguments("Buffer10Container.this[int]", "Buffer10").WithLocation(2, 31)
+                );
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/67533")]
+        public void InObjectInitializer_Range()
+        {
+            string source = """
+Buffer10 b = default;
+b[..] = null; // 1
+
+_ = new Buffer10() { [..] = null }; // 2
```

Comment:
>null [](http://example.com/codeflow?start=28&length=4)

Nested initializer should probably work with the slice scenario. Are we testing it?